File: 2/python/part2.py
import sys

horizontal_position = 0
depth = 0
aim = 0

for line in sys.stdin:
    instruction, distance = line.rstrip().split(' ')

    if instruction == "forward":
        horizontal_position += int(distance)
        depth += aim * int(distance)
    elif instruction == "down":
        aim += int(distance)
    elif instruction == "up":
        aim -= int(distance)

print(horizontal_position)
print(depth)
print(horizontal_position * depth)
    

File: 2/js/part2.js
const fs = require('fs');

const instructionToMotions = {
  'forward': (x, h, d, a) => ([h + x, d + a * x, a]),
  'up': (x, h, d, a) => ([h, d, a - x]),
  'down': (x, h, d, a) => ([h, d, a + x]),
};

fs.readFile(process.argv[2], 'utf-8', (err, data) => {
  console.log(
    data.split("\n")
      .map(s => s.split(' '))
      .map(([instruction, s]) => ([instruction, parseInt(s)]))
      .reduce(([h, d, a], [instruction, x]) => instructionToMotions[instruction](x, h, d, a), [0, 0, 0])
      .slice(0, 2)
      .reduce((acc, n) => acc * n, 1)
  );
});
